shumai/module/transformer.ts

```diff
@@ -451,3 +460,172 @@ export class TransformerEncoder extends Module {
+  private attentionDim: number
+  private feedForwardDim: number
+  private maskedMhsa: TransformerMultiheadAttention
+  private maskedMhsaNorm: LayerNorm
+  private mhca: TransformerMultiheadAttention
+  private mhcaNorm: LayerNorm
+  private ff: FeedForward
+  private ffNorm: LayerNorm
+
+  /**
+   * @param dim - Number of dimensions of the input embeddings
+   * @param heads - Number of heads for the multi-head attention
+   * @param attentionDim - Number of dimensions of the embeddings used in the scaled dot-product attention, or `dim` if not specified
+   * @param feedForwardDim - Number of dimensions in the hidden layer of the feed forward network, or `dim` if not specified
+   */
+  constructor(dim: number, heads: number, attentionDim?: number, feedForwardDim?: number) {
+    super()
+
+    this.dim = dim
+    this.heads = heads
+    if (attentionDim === undefined) {
+      this.attentionDim = dim
+    } else {
+      this.attentionDim = attentionDim
+    }
+    if (feedForwardDim === undefined) {
+      this.feedForwardDim = dim
+    } else {
+      this.feedForwardDim = feedForwardDim
+    }
+
+    this.maskedMhsa = new TransformerMultiheadAttention(this.dim, this.heads, this.attentionDim)
+    this.maskedMhsaNorm = new LayerNorm([this.dim])
+    this.mhca = new TransformerMultiheadAttention(this.dim, this.heads, this.attentionDim)
+    this.mhcaNorm = new LayerNorm([this.dim])
+    this.ff = new FeedForward(this.dim, this.feedForwardDim)
+    this.ffNorm = new LayerNorm([this.dim])
+  }
+
+  /**
+   * @param sequenceLength - Length of sequence for which the mask should be generated
+   * @returns A Tensor mask of shape `[sequenceLength, sequenceLength]` where row $i$ should have 0s in positions up to $i$ and 1s everywhere else
+   */
+  static getSelfAttentionMask(sequenceLength: number): Tensor {
+    return sm
+      .full([sequenceLength, sequenceLength], 1)
+      .astype(sm.dtype.BoolInt8)
+      .tril()
+      .logicalNot()
+  }
+
+  /**
+   * @param input - Tensor from the previous decoder layer, shape `[..., tokens, dim]`
+   * @param encoderOutput - Tensor output by the encoder, shape `[..., encoderTokens, dim]`
+   * @returns A Tensor of shape `[..., tokens, dim]`
+   */
+  forward(input: Tensor, encoderOutput: Tensor): Tensor {
+
+    const decoderLength = input.shape[input.shape.length - 2]
+    const mask = TransformerDecoderLayer.getSelfAttentionMask(decoderLength)
+
+    let residual = input
+    let output = this.maskedMhsa(input, input, input, mask) // shape [..., tokens, dim]
+    output = this.maskedMhsaNorm(residual.add(output))
+
+    residual = output
+    output = this.mhca(output, encoderOutput, encoderOutput) // shape [..., tokens, dim]
+    output = this.mhcaNorm(residual.add(output))
+
+    residual = output
+    output = this.ff(output) // shape [..., tokens, dim]
+    output = this.ffNorm(residual.add(output))
+
+    return output
+  }
+}
+
+class StackableDecoderLayer extends Module {
```

Comment:
If we relax the type of `Sequential` to use lists of `CallableFunction`, we could probably avoid creating this entirely.
```javascript
       const td = new TransformerDecoderLayer(this.dim, this.heads, this.attentionDim, this.feedForwardDim)
       layers.push(
         (input, encoderOutput) => [ td(input, encoderOutput), encoderOutput ]
       )
    // ...
    this.layers = new Sequential(...layers)
```


This is because Shumai has the nice property (not in TF/PT) that parameters are tracked regardless of being in a Module.  In fact, the module concept can be implemented entirely with captures (rather than classes).

Reply:
Ah! I thought that there would be a more idiomatic way of doing this, but couldn't think of one off the top of my head. Thank you for the suggestion!